core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -286,13 +287,23 @@ class ReplicaFetcherThread(name: String,
   }
 
   /**
-    * - Truncate the log to the leader's offset for each partition's epoch.
-    * - If the leader's offset is greater, we stick with the Log End Offset
-    *   otherwise we truncate to the leaders offset.
-    * - If the leader replied with undefined epoch offset we must use the high watermark
-    */
-  override def maybeTruncate(fetchedEpochs: Map[TopicPartition, EpochEndOffset]): ResultWithPartitions[Map[TopicPartition, Long]] = {
-    val fetchOffsets = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+   * Truncate the log for each partition's epoch based on leader's returned epoch and offset.
+   *  -- If the leader replied with undefined epoch offset, we must use the watermark. This can
+   *  happen if 1) the leader is on the protocol version < KAFKA_0_11_0_IV2; 2) the follower
+   *  requested leader epoch < the first leader epoch known to the leader.
+   *  -- If the leader replied with the valid offset but undefined leader epoch, we truncate to
+   *  leader's offset if it is lower than follower's Log End Offset. This may happen if the
+   *  leader in on the protocol version < KAFKA_1_1_IV0
+   *  -- If the leader replied with leader epoch not known to the follower, we truncate to the
+   *  end offset of the largest epoch that is smaller than the epoch the leader replied with, and
+   *  send offset for leader epoch request with that leader epoch. In a more rare case, where the
+   *  follower was not tracking epochs smaller than the epoch the leader replied with, we
+   *  truncate the leader's offset (and do not send any more leader epoch requesrs).
```

Comment:
My concern about falling back to high watermark in this particular case is that post-KIP-101 code (and pre-2.0) behaves exactly as described, since the leader does not send leader epoch, so we don't check it and use leader's offset to truncate. And also if brokers upgrade to 2.0, but do not upgrade protocol version. Then, we upgrade to 2.0 protocol version, and we are back to high watermark in this case.